ayeletK/assignment2/Q2_solution.cpp

```diff
@@ -0,0 +1,71 @@
+#include "tree_node.cpp"
+#include <string>
+
+using namespace std;
+
+#define KEY_NOT_IN_PATH "-"
+#define EMPTY_PATH ""
+#define LEFT_SON '0'
+#define RIGHT_SON '1'
+
+template <typename Key>
+char get_addition_to_path(Node<Key>* son_node, Node<Key>* parent_node){
+  return parent_node->left == son_node ? LEFT_SON : RIGHT_SON;
+}
+
+template <typename Key>
+string get_path_string(Key key, Node<Key>* curr_node, Node<Key>* parent_node = NULL){
+  if (!curr_node) return KEY_NOT_IN_PATH;
+  Key curr_key = curr_node->get_key();
+  // EMPTY_PATH represents the case the root node is the key
+  if (curr_key == key){
+    if (!parent_node) return EMPTY_PATH;
+    else return string(1, get_addition_to_path(curr_node, parent_node));
+  }
+  string res;
+  res = get_path_string(key, curr_node->left, curr_node);
+  if (res == KEY_NOT_IN_PATH){
+    res = get_path_string(key, curr_node->right, curr_node);
+  }
+  if (res != KEY_NOT_IN_PATH && parent_node){
+    char addition = get_addition_to_path(curr_node, parent_node);
```

Comment:
I think you can omit the parent node parameter. Instead of "peeking" to the parent, you can use the fact that you know whether you called the left or right node.

```C++
// Line 25 and forward:
string res;
res = get_path_string(key, curr_node->left);  // No parent node, simpler function signature.
if (res != KEY_NOT_IN_PATH) {
    res.insert(res.begin(), LEFT_SON);
    return res;
}

res = get_path_string(key, curr_node->right);
if (res != KEY_NOT_IN_PATH) {
    res.insert(res.begin(), RIGHT_SON);
    return res;
}
return res;  // Key not in subtree.
```

Reply:
let the parent concat the correct addition to the result string?
Doesn't it considered a code duplication? (you said you were hunting for those so I tries to avoid it as much as possible)

Reply:
The DRY principle says:
 "Every piece of knowledge must have a single, unambiguous, authoritative representation within a system"
As I understand it, it means that constant values should be defined in one place, and specific calculations should not repeat, etc.
So I 'hunt' for *ideas* that repeat, and not necessarily *text* that repeats in the code.

For example, calling a single function several times in a row is a "text" duplication, but not a duplication of the idea (which is already extracted into the function!).

In the example code above, I called the same code on the left and right nodes. I run `get_path_string` and check its result. Yes, I write almost the same identical code twice, but the core of it is already extracted mostly to functions, so I don't see this as a bad duplication. I think binary trees will always tend to a little of duplication when each recursive call is made on both the left and right child.

... This is all very subjective and reflects my personal opinion - the code above is a suggestion on how to simplify the code (by removing the `parent_node` parameter). Our goal is simple, readable code. Avoiding duplication is just one aspect of it :) Sometimes programming principles can contradict each other, our goal as coders is to try to find the best tradeoff.

--------------
BUT if you insist, I we can avoid some duplication in the above code ;)
```C++
string res;
for (Side side : {LEFT_SON, RIGHT_SON}) {
    Node* child_node = (side == LEFT_SON) ? curr_node->left : curr_node->right;
    res = get_path_string(key, child_node);
    if (res != KEY_NOT_IN_PATH) {
        res.insert(res.begin(), side);
        return res;
    }
}
return res;  // Key not in subtree.
```
This is the same code with less duplication. I don't think this is necessary more readable, I guess that depends on how this pattern (of 'iterating' on both children) is common in the code. (while calling the left and right child sequentially is common for sure when handling binary trees).